File: app/rag/chunker.py
from __future__ import annotations

import re
from app.config import AppConfig
from app.models import TextChunk


_SENTENCE_END = re.compile(r"[.!?]\s+")
_NEWLINE      = re.compile(r"\n")


class TextChunker:
    def __init__(self, cfg: AppConfig) -> None:
        self._max_size = cfg.chunking.max_chunk_size
        self._overlap  = cfg.chunking.overlap

    def chunk(self, text: str) -> list[TextChunk]:
        """
        Split text into overlapping chunks no larger than max_size characters.

        Break priority:
          1. Last sentence boundary ([.!?] followed by whitespace) inside the window
          2. Last newline inside the window — natural for code where sentence breaks are rare
          3. Hard character cut — absolute fallback so a chunk is never > max_size chars

        This ensures code with long methods never produces a chunk that exceeds the
        configured size, regardless of whether natural sentence boundaries exist.
        """
        if not text:
            return []

        # Fast path: fits in one chunk
        if len(text) <= self._max_size:
            return [TextChunk(text=text, index=0, start_offset=0, end_offset=len(text))]

        chunks: list[TextChunk] = []
        start = 0
        idx   = 0

        while start < len(text):
            end = min(start + self._max_size, len(text))

            if end < len(text):
                window = text[start:end]

                # Priority 1: last sentence boundary
                best_break = None
                for m in _SENTENCE_END.finditer(window):
                    best_break = m.end()

                # Priority 2: last newline (crucial for code — keeps lines intact)
                if best_break is None:
                    for m in _NEWLINE.finditer(window):
                        best_break = m.end()

                # Reject a break that's too close to the start (< 25% of window)
                # to avoid tiny slivers that stall progress.
                min_break = max(1, self._max_size // 4)
                if best_break is not None and best_break < min_break:
                    best_break = None  # fall through to hard cut

                if best_break is not None:
                    end = start + best_break

            chunk_text = text[start:end].strip()
            if chunk_text:
                chunks.append(TextChunk(
                    text=chunk_text,
                    index=idx,
                    start_offset=start,
                    end_offset=end,
                ))
                idx += 1

            if end >= len(text):
                break

            # Advance with overlap — always move forward by at least 1 char
            start = max(start + 1, end - self._overlap)

        return chunks

File: internal/rag/chunker.go
// Text chunker — Go port of app/rag/chunker.py.
package rag

import (
	"log"
	"regexp"
	"strings"

	"github.com/MichalOndrejka/conduit/internal/config"
	"github.com/MichalOndrejka/conduit/internal/models"
)

var (
	sentenceEnd = regexp.MustCompile(`[.!?]\s+`)
	newline     = regexp.MustCompile(`\n`)
)

type TextChunker struct {
	maxSize int
	overlap int
}

func NewTextChunker(cfg *config.AppConfig) *TextChunker {
	c := &TextChunker{
		maxSize: cfg.Chunking.MaxChunkSize,
		overlap: cfg.Chunking.Overlap,
	}
	effectiveEmbedChars := cfg.Embedding.MaxInputTokens * charsPerToken
	if c.maxSize > effectiveEmbedChars {
		log.Printf(
			"warning: max_chunk_size (%d chars) exceeds embedding model limit (%d tokens × %d chars/token = %d chars) — clamping chunk size",
			c.maxSize, cfg.Embedding.MaxInputTokens, charsPerToken, effectiveEmbedChars,
		)
		c.maxSize = effectiveEmbedChars
	}
	return c
}

// Chunk splits text into overlapping chunks no larger than maxSize characters.
//
// Break priority (same as the Python implementation):
//  1. Last sentence boundary ([.!?] followed by whitespace) inside the window
//  2. Last newline inside the window — natural for code
//  3. Hard character cut — absolute fallback
func (c *TextChunker) Chunk(text string) []models.TextChunk {
	if text == "" {
		return nil
	}

	// Fast path: fits in one chunk
	if len(text) <= c.maxSize {
		return []models.TextChunk{{Text: text, Index: 0, StartOffset: 0, EndOffset: len(text)}}
	}

	var chunks []models.TextChunk
	start := 0
	idx := 0

	for start < len(text) {
		end := min(start+c.maxSize, len(text))

		if end < len(text) {
			window := text[start:end]

			// Priority 1: last sentence boundary
			bestBreak := -1
			for _, m := range sentenceEnd.FindAllStringIndex(window, -1) {
				bestBreak = m[1]
			}

			// Priority 2: last newline (crucial for code — keeps lines intact)
			if bestBreak < 0 {
				for _, m := range newline.FindAllStringIndex(window, -1) {
					bestBreak = m[1]
				}
			}

			// Reject a break too close to the start (< 25% of window) to
			// avoid tiny slivers that stall progress.
			minBreak := max(1, c.maxSize/4)
			if bestBreak >= 0 && bestBreak < minBreak {
				bestBreak = -1 // fall through to hard cut
			}

			if bestBreak >= 0 {
				end = start + bestBreak
			}
		}

		chunkText := strings.TrimSpace(text[start:end])
		if chunkText != "" {
			chunks = append(chunks, models.TextChunk{
				Text:        chunkText,
				Index:       idx,
				StartOffset: start,
				EndOffset:   end,
			})
			idx++
		}

		if end >= len(text) {
			break
		}

		// Advance with overlap — always move forward by at least 1 char
		start = max(start+1, end-c.overlap)
	}

	return chunks
}
